akka-stream/src/main/scala/akka/stream/impl/io/compression/DeflateCompressor.scala

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.impl.io.compression
+
+import java.util.zip.Deflater
+
+import akka.util.{ ByteString, ByteStringBuilder }
+
+import scala.annotation.tailrec
+
+/** INTERNAL API */
+private[akka] class DeflateCompressor extends Compressor {
+  import DeflateCompressor._
+
+  protected lazy val deflater = new Deflater(Deflater.BEST_COMPRESSION, false)
+
+  override final def compressAndFlush(input: ByteString): ByteString = {
+    val buffer = newTempBuffer(input.size)
+
+    compressWithBuffer(input, buffer) ++ flushWithBuffer(buffer)
+  }
+  override final def compressAndFinish(input: ByteString): ByteString = {
+    val buffer = newTempBuffer(input.size)
+
+    compressWithBuffer(input, buffer) ++ finishWithBuffer(buffer)
+  }
+  override final def compress(input: ByteString): ByteString = compressWithBuffer(input, newTempBuffer())
+  override final def flush(): ByteString = flushWithBuffer(newTempBuffer())
+  override final def finish(): ByteString = finishWithBuffer(newTempBuffer())
+
+  protected def compressWithBuffer(input: ByteString, buffer: Array[Byte]): ByteString = {
+    require(deflater.needsInput())
+    deflater.setInput(input.toArray)
+    drainDeflater(deflater, buffer)
+  }
+  protected def flushWithBuffer(buffer: Array[Byte]): ByteString = {
+    val written = deflater.deflate(buffer, 0, buffer.length, Deflater.SYNC_FLUSH)
+    ByteString.fromArray(buffer, 0, written)
+  }
+  protected def finishWithBuffer(buffer: Array[Byte]): ByteString = {
+    deflater.finish()
+    val res = drainDeflater(deflater, buffer)
+    deflater.end()
```

Comment:
I think it's important to always invoke `end()` for the Inflater and/or Deflater because it/they allocate off-heap memory. Add that to `postStop` of the stage?